qiskit/providers/aer/aer_snapshots.py

```diff
@@ -0,0 +1,332 @@
+        for tuple_element in tuples:
+            if isinstance(tuple_element, QuantumRegister):
+                for j in range(tuple_element.size):
+                    qubits.append(tuple_element[j])
+            else:
+                qubits.append(tuple_element)
+    return self.append(
+        Snapshot(
+            label,
+            snapshot_type=snapshot_type,
+            num_qubits=len(qubits),
+            params=params), qubits)
+
+
+# Add to QuantumCircuit class
+QuantumCircuit.snapshot = snapshot
+
+class SnapshotStatevector(Snapshot):
+    """ Simulator snapshot instruction for statevector snapshot type """
+
+    def __init__(self,
+                 label,
+                 snapshot_type='statevector',
+                 num_qubits=0,
+                 num_clbits=0,
+                 params=None):
+        super().__init__(label, snapshot_type, num_qubits, num_clbits, params)
+
+
+def snapshot_statevector(self,
+                         label,
+                         snapshot_type='statevector',
+                         qubits=None,
+                         params=None):
+
+    # Convert label to string for backwards compatibility
+    if not isinstance(label, str):
+        warnings.warn(
+            "Snapshot label should be a string, "
+            "implicit conversion is deprecated.", DeprecationWarning)
+        label = str(label)
+    # If no qubits are specified we add all qubits so it acts as a barrier
+    # This is needed for full register snapshots
+    if isinstance(qubits, QuantumRegister):
+        qubits = qubits[:]
+    if not qubits:
+        tuples = []
+        if isinstance(self, QuantumCircuit):
+            for register in self.qregs:
+                tuples.append(register)
+        if not tuples:
+            raise ExtensionError('no qubits for snapshot')
+        qubits = []
+        for tuple_element in tuples:
+            if isinstance(tuple_element, QuantumRegister):
+                for j in range(tuple_element.size):
+                    qubits.append(tuple_element[j])
+            else:
+                qubits.append(tuple_element)
+    return self.append(
+        SnapshotStatevector(
+            label,
+            snapshot_type=snapshot_type,
+            num_qubits=len(qubits),
+            params=params), qubits)
+
+QuantumCircuit.snapshot_statevector = snapshot_statevector
+
+
+class SnapshotStabilizer(Snapshot):
+    def __init__(self,
+                 label,
+                 snapshot_type='stabilizer',
+                 num_qubits=0,
+                 num_clbits=0,
+                 params=None):
+        super().__init__(label, snapshot_type, num_qubits, num_clbits, params)
+
+def snapshot_stabilizer(self,
+                        label,
+                        snapshot_type='stabilizer',
+                        qubits=None,
+                        params=None):
+
+    # Convert label to string for backwards compatibility
+    if not isinstance(label, str):
```

Comment:
Could all of the first part of this method (up to the return) be extracted into a separate method? That way we reduce the duplication(and so improve the maintainability of the code) as all the different types of snapshot can call on that one method instead of having their own copy of the code. It could be made a private method of `Snapshot`.

Reply:
Done! Thanks for the help